cmd/dex/serve.go

```diff
@@ -68,6 +71,28 @@ func commandServe() *cobra.Command {
 	return cmd
 }
 
+func listenAndShutdownGracefully(logger log.Logger, gr *run.Group, srv *http.Server, name string) error {
+	l, err := net.Listen("tcp", srv.Addr)
+	if err != nil {
+		return fmt.Errorf("listening (%s) on %s: %v", name, srv.Addr, err)
+	}
+
+	gr.Add(func() error {
+		logger.Infof("listening (%s) on %s", name, srv.Addr)
+		return srv.Serve(l)
+	}, func(err error) {
+		logger.Debugf("starting gracefully shutdown (%s)", name)
+		if err := l.Close(); err != nil {
+			logger.Errorf("gracefully close (%s) listener: %v", name, err)
+		}
```

Comment:
AFAIK the http server takes care of this, so you don't need to call close on the listener manually.

Reply:
My thought was about splitting listener closing error and server shutdown error. Maybe it is excessive.

Reply:
Listener closing now is a duty of a server.